### PR TITLE
Canvas fingerprinting and supercookies

### PR DESCRIPTION
**Pull request by cooperq**
Working canvas fingerprinting detection.
To test: go to expedia.com and note that www.cdn-net.com will show up in the popup, and there will be a console.log message about cdn-net fingerprinting on expedia.com
note that cdn-net does not show up in the popup on master branch. 

cc @SwartzCr for review

### CODE REVIEW

**Comment by ghostwords** [comment]
This might be missing the "interpret data" bit: take a look at [recordFingerprinting()](https://github.com/EFForg/privacybadgerchrome/blob/55a639a1d3d6c3d8b632f5df70c98d335e05044e/src/webrequest.js#L294-L317) in PBChrome.

Also, should [ignore scriptUrls when we fail to get them](https://github.com/EFForg/privacybadgerchrome/commit/b35188df3cad0655cb666b5512b6cae508c3578d), and should disable pageMod injection on whitelisted pages/when PB is globally disabled.

**Comment by cooperq** [comment]
@ghostwords we are checking that privacy badger is enabled on the domain before we mark the third party as fingerprinting, but I think there is no good way for us to exclude the content script from certain domains, there is the exclude field but that only gets checked when page mod gets called which only happens when the addon is loaded.

**Comment by cooperq** [comment]
Okay the last commit implements the `recordFingerprinting()` function from chrome.

**Comment by cooperq** [comment]
supercookies can be tested at: https://cooperq.com/supercookie.html (rawgit)
fingerprinting can be tested at: https://cooperq.com/canvas.html (beaverbird)

File: lib/heuristicBlocker.js [hunk 1]
```diff
@@ -4,18 +4,20 @@
...
-const { on, once, off, emit } = require('sdk/event/core');
+const { emit } = require('sdk/event/core');
 const { settingsMap } = require("./ui");
 const {
   checkPrivacyPolicy,
   whitelistHostFromPrivacyPolicy
 } = require("./policyCheck");
 const { doDependingOnIsPrivate } = require("./privateBrowsing");
+const prefs = require('sdk/simple-prefs').prefs;
+const pbUI = require("./ui");
```
Comment (SwartzCr) [inline lib/heuristicBlocker.js]: duplicate

File: lib/heuristicBlocker.js [hunk 2]
```diff
@@ -331,12 +333,31 @@ let getChannelInfo = function(channel, win) {
  * @return Boolean Whether this request appears to be tracking the user
  */
 let hasTracking = function(channel, channelInfo, cookies) {
+  let tracking = hasCookieTracking(channel, channelInfo, cookies) ||
+    hasSupercookieTracking(channel, channelInfo) ||
+    hasCanvasFingerprinting(channel, channelInfo);
 
+  return tracking;
```
Comment (SwartzCr) [inline lib/heuristicBlocker.js]: can we just return this instead of let and then return?

File: lib/heuristicBlocker.js [hunk 3]
```diff
@@ -331,12 +333,31 @@ let getChannelInfo = function(channel, win) {
...
+    settingsMap.canvasTracking[channelInfo.origin] &&
+    !!settingsMap.canvasTracking[channelInfo.origin].fingerprinting;
+};
+
+var hasCookieTracking = function(channel, channelInfo, cookies){
   if (!cookies || Object.getOwnPropertyNames(cookies).length < 1) {
     return false;
   }
 
-  let { origin, parentOrigin } = channelInfo;
```
Comment (SwartzCr) [inline lib/heuristicBlocker.js]: do we really not need this? we're still referring to it later

File: lib/main.js [hunk 4]
```diff
@@ -1,27 +1,44 @@
...
-const windows = require('sdk/windows').browserWindows;
-const socialWidgetHandler = require("./socialWidgetHandler");
 const progressListener = require("./progressListener");
+const socialWidgetHandler = require("./socialWidgetHandler");
+const { storage } = require("sdk/simple-storage");
+const tabs = require("sdk/tabs");
+const userStorage = require("./userStorage");
+const utils = require("./utils");
+const windows = require('sdk/windows').browserWindows;
+const { settingsMap } = require("./ui");
```
Comment (SwartzCr) [inline lib/main.js]: duplicate?

File: lib/heuristicBlocker.js (hunk 3 above)
Comment (cooperq) [inline lib/heuristicBlocker.js]: I only removed getting parentOrigin since we don't use it in this function, we are still getting origin

**Comment by cooperq** [comment]
@SwartzCr this is ready for testing now, I have fixed the style nits.

**Comment by SwartzCr** [comment]
looks to be working!

**Comment by antistress** [comment]
Hi,

Note that supercookies counter-mesure is not mentionned on the FAQ (whereas canvas fingerprinting is)

**Comment by cooperq** [comment]
Thanks, we will update the FAQ!

**Pull request merged**